finality-aleph/src/sync/handler.rs

```diff
@@ -105,19 +110,59 @@ impl<I: PeerId, J: Justification, CS: ChainStatus<J>, V: Verifier<J>, F: Finaliz
         finalizer: F,
         period: SessionPeriod,
     ) -> Result<Self, Error<J, CS, V, F>> {
-        let top_finalized = chain_status
-            .top_finalized()
-            .map_err(Error::ChainStatus)?
-            .header()
-            .id();
-        let forest = Forest::new(top_finalized);
-        Ok(Handler {
+        let forest = Forest::new(
+            chain_status
+                .top_finalized()
+                .map_err(Error::ChainStatus)?
+                .header()
+                .id(),
+        );
+        let mut handler = Handler {
             chain_status,
             verifier,
             finalizer,
             forest,
             session_info: SessionBoundaryInfo::new(period),
-        })
+        };
+        handler.refresh_forest()?;
+        Ok(handler)
+    }
+
+    /// TODO: Remove after completing the sync rewrite.
+    /// Move the code to `Self::new` to initialize the `Forest` properly.
+    pub fn refresh_forest(&mut self) -> Result<(), Error<J, CS, V, F>> {
+        let top_finalized = self
+            .chain_status
+            .top_finalized()
+            .map_err(Error::ChainStatus)?
+            .header()
+            .id();
+        let mut forest = Forest::new(top_finalized.clone());
+        let mut deque = VecDeque::from([top_finalized]);
+        while let Some(hash) = deque.pop_front() {
+            let children = self
+                .chain_status
+                .children(hash)
+                .map_err(Error::ChainStatus)?;
+            for header in children.iter() {
+                if let Err(e) = forest.update_body(header) {
+                    match e {
+                        ForestError::TooNew => {
+                            warn!(
+                                target: LOG_TARGET,
+                                "There are more imported non-finalized blocks that can fit into the Forest: {}.", e
+                            );
+                            self.forest = forest;
+                            return Ok(());
+                        }
+                        _ => return Err(Error::Forest(e)),
+                    }
```

Comment:
I was thinking of smth like this
```rust
match forest.update_body(header) {
   Err(ForestError::TooNew) => ...
   Err(e) => ...,
   _ => ...,
}
```
wdyt?

Reply:
the last branch would be empty, right? I think it's better to leave `if let Err(e)`

Reply:
so be it